### PR TITLE
feat(azure): implement mergePr

### PR DESCRIPTION
**Pull request by elwynelwyn**
## Changes:

Implement the `mergePr` function for Azure DevOps. This updates a PR, setting the status to `Completed` and passing the required `lastMergeSourceCommit`.

This allows automerge to work end-to-end on Azure.

## Context:

#1149 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [X] Unit tests + ran on a real repository

Ran locally and pointed at my works AzureDevops repo.

### CODE REVIEW

**Comment by viceice** [comment]
@elwynelwyn next time simply update git commit and force push, you don't need to open a new pr to fix git author. 😉

File: lib/platform/azure/index.ts
Comment (viceice) [inline lib/platform/azure/index.ts:606]:
```suggestion
  return true;
```

**Review by viceice — CHANGES_REQUESTED** [review]
There should be a try catch and return falls on failure.

**Review by viceice — DISMISSED** [review]
@JamieMagee Can you please verify this?

**Comment by JamieMagee** [comment]
@elwynelwyn This doesn't work for me. I get the error message:

> Unable to complete the pull request. The chosen merge type is forbidden by policy.

You need to explicitly set a merge method. Please take a look at how we [set autocomplete][1] and how we [get allowed merge methods][2]

[1]: https://github.com/renovatebot/renovate/blob/master/lib/platform/azure/index.ts#L384-L397
[2]: https://github.com/renovatebot/renovate/blob/f18b7f3504c3bdb799f1ff15fbe977cefb24641d/lib/platform/azure/azure-helper.ts#L178-L199

**Review by JamieMagee — CHANGES_REQUESTED** [review]
See comment

**Comment by elwynelwyn** [comment]
> @elwynelwyn This doesn't work for me. I get the error message:
> 
> > Unable to complete the pull request. The chosen merge type is forbidden by policy.
> 
> You need to explicitly set a merge method. Please take a look at how we [set autocomplete](https://github.com/renovatebot/renovate/blob/master/lib/platform/azure/index.ts#L384-L397) and how we [get allowed merge methods](https://github.com/renovatebot/renovate/blob/f18b7f3504c3bdb799f1ff15fbe977cefb24641d/lib/platform/azure/azure-helper.ts#L178-L199)

I see `config.mergeMethod` is based on just the repoId + project, however it's possible for policies to be set per branch, but the first policy is always selected (I hit this when I added a policy to test this change - the first policy, for `master` was selected, however the 2nd test policy I setup on my renovatebot baseBranch was not selected, so my merge failed with the same error you got ^^). 

So I extended `getMergeMethod`  and call it each time from `mergePr` rather than using the existing `config.mergeMethod` value - that way the `pr.targetRef` is available to get the most specific policy.

Perhaps it wouldn't be an issue if we used renovate's `baseBranch` value initially when building the config, but I'm not sure of the best way of getting that value into `initRepo` where the config is first built (looks like perhaps RenovateConfig could be passed [all the way down from here](https://github.com/renovatebot/renovate/blob/69931df927ed6c62f54a5c368259be112067299f/lib/workers/repository/init/apis.ts#L60)? not sure how complete the renovate config is at that point). I wonder if other places config.mergeMethod is used could hit this scenario.

Let me know if you'd prefer me to revert all this and just simply use `config.mergeMethod` on the PR update and not worry about the branch policy stuff now :)

**Comment by viceice** [comment]
I'm ok to fetch this info on mergePr, but we should cache this info(on `config`, so it will reset on repo change.

**Comment by JamieMagee** [comment]
I hit this same issue before. See https://github.com/renovatebot/renovate/issues/7310#issuecomment-731762556.

I think the `git/policy/configurations` is the correct API to use, rather than `policy/configurations`. However we're still blocked by a lack of an update API release (check out the linked comment for a more in depth explanation).

Your solutions looks to be a better fix in the meantime. One small improvement I can see, is using the `matchKind` from the policy scope. It can be `Prefix` or `Exact`. So we could do something like `s.matchKind === 'Exact' ? s.refName === branchRef : branchRef.startsWith(s.refName)`. What do you think?

**Review by JamieMagee — CHANGES_REQUESTED** [review]
See comments.

I reviewed this and forgot to set my status. Apologies.

**Comment by elwynelwyn** [comment]
> we should cache this info(on `config`, so it will reset on repo change.

Can you please point me somewhere we do this already? Or do you mean to just assign the new `mergeMethod` back onto the `config` object? (I'm not 100% sure how on subsequent merges we'd then check the existing value as it would be unclear if it was set based on a `branchRef` or not)

> One small improvement I can see, is using the `matchKind` from the policy scope. It can be `Prefix` or `Exact`. So we could do something like `s.matchKind === 'Exact' ? s.refName === branchRef : branchRef.startsWith(s.refName)`. What do you think?

This is done, and I pulled the `some` out into a function as it was getting a bit nested and unclear.

File: lib/platform/azure/index.ts
Comment (viceice) [inline lib/platform/azure/index.ts:593]: Some caching, needs `config` type changes:

```suggestion
  const mergeMethod = config.mergeMethods[pr.targetRefName] ?? (config.mergeMethods[pr.targetRefName] = await azureHelper.getMergeMethod(
    config.repoId,
    config.project,
    pr.targetRefName
  ));
```

This is what i had in mind with caching the mergeMethods results.

**Comment by elwynelwyn** [comment]
Thanks @viceice, I have made that change now.

Also renamed the existing `config.mergeMethod` to `config.defaultMergeMethod` (got a better name for me?) as it seems a little confusing having both `mergeMethod`/`mergeMethods`. Could also just have the single object of merge methods and put the default one on it with a special non-branch-ref key or something. Let me know what you think

**Review by viceice — APPROVED** [review]
Waiting for @JamieMagee 's approval

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**
